File: experiments-local/experiment-19/runs/agent=omp_language=python_model=claude-sonnet-4-6_tooling=none/rep2/query_engine.py
"""
Query engine for Brazilian Soccer MCP Server.

All query logic lives here, independent of the MCP transport layer,
making it straightforwardly testable.
"""

from typing import Any, Optional

import pandas as pd

from data_loader import SoccerDataLoader, normalize_team, strip_accents

COMPETITION_ALIASES: dict[str, list[str]] = {
    "brasileirao": ["brasileirão", "brasileirao", "serie a", "campeonato brasileiro"],
    "copa do brasil": ["copa do brasil", "copa brasil", "brazilian cup"],
    "copa libertadores": ["libertadores", "copa libertadores"],
}


def _resolve_competition(query: str) -> str:
    """Return accent-stripped canonical competition search term."""
    q = strip_accents(query.lower())
    for canon, aliases in COMPETITION_ALIASES.items():
        for alias in aliases:
            if strip_accents(alias) in q:
                return strip_accents(canon)
    return q


def _comp_mask(series: pd.Series, comp: str) -> pd.Series:
    """Accent-insensitive competition contains check."""
    comp_norm = strip_accents(comp)
    return series.apply(lambda x: strip_accents(str(x).lower())).str.contains(
        comp_norm, na=False
    )


class QueryEngine:
    """
    Executes queries against the loaded soccer data.

    All methods return plain Python dicts/lists that the MCP server
    formats into human-readable strings before returning to callers.
    """

    def __init__(self, loader: Optional[SoccerDataLoader] = None):
        self.loader = loader or SoccerDataLoader()

    # ------------------------------------------------------------------ #
    # Match queries                                                        #
    # ------------------------------------------------------------------ #

    def search_matches(
        self,
        team: str,
        opponent: Optional[str] = None,
        competition: Optional[str] = None,
        season: Optional[int] = None,
        date_from: Optional[str] = None,
        date_to: Optional[str] = None,
        limit: int = 50,
    ) -> list[dict]:
        """
        Return matches involving *team*, optionally filtered by opponent,
        competition, season, or date range.  Results are sorted newest-first.
        """
        df = self.loader.matches
        norm = normalize_team(team)

        mask = df["home_team"].str.contains(norm, na=False) | df["away_team"].str.contains(
            norm, na=False
        )

        if opponent:
            norm_opp = normalize_team(opponent)
            mask &= df["home_team"].str.contains(norm_opp, na=False) | df[
                "away_team"
            ].str.contains(norm_opp, na=False)

        if competition:
            comp = _resolve_competition(competition)
            mask &= _comp_mask(df["competition"], comp)

        if season is not None:
            mask &= df["season"] == int(season)

        if date_from:
            mask &= df["date"] >= date_from

        if date_to:
            mask &= df["date"] <= date_to

        result = df[mask].sort_values("date", ascending=False).head(limit)
        return result.to_dict("records")

    def head_to_head(
        self,
        team1: str,
        team2: str,
        competition: Optional[str] = None,
        season: Optional[int] = None,
    ) -> dict:
        """
        Return head-to-head record between two teams plus match list.
        """
        matches = self.search_matches(
            team=team1,
            opponent=team2,
            competition=competition,
            season=season,
            limit=500,
        )

        norm1 = normalize_team(team1)
        norm2 = normalize_team(team2)

        t1_wins = t2_wins = draws = 0
        for m in matches:
            ht, at = m["home_team"], m["away_team"]
            hg, ag = m["home_goal"], m["away_goal"]
            if norm1 in ht and norm2 in at:
                if hg > ag:
                    t1_wins += 1
                elif ag > hg:
                    t2_wins += 1
                else:
                    draws += 1
            elif norm2 in ht and norm1 in at:
                if ag > hg:
                    t1_wins += 1
                elif hg > ag:
                    t2_wins += 1
                else:
                    draws += 1
            else:
                # Fallback: team1 substring match
                if hg > ag:
                    t1_wins += 1
                elif ag > hg:
                    t2_wins += 1
                else:
                    draws += 1

        return {
            "team1": team1,
            "team2": team2,
            "team1_wins": t1_wins,
            "team2_wins": t2_wins,
            "draws": draws,
            "total_matches": len(matches),
            "matches": matches,
        }

    # ------------------------------------------------------------------ #
    # Team statistics                                                      #
    # ------------------------------------------------------------------ #

    def get_team_stats(
        self,
        team: str,
        competition: Optional[str] = None,
        season: Optional[int] = None,
        venue: str = "both",  # "home" | "away" | "both"
    ) -> dict:
        """Return wins/losses/draws/goals for a team."""
        df = self.loader.matches
        norm = normalize_team(team)

        home_mask = df["home_team"].str.contains(norm, na=False)
        away_mask = df["away_team"].str.contains(norm, na=False)

        if competition:
            comp = _resolve_competition(competition)
            comp_mask = _comp_mask(df["competition"], comp)
            home_mask &= comp_mask
            away_mask &= comp_mask

        if season is not None:
            season_mask = df["season"] == int(season)
            home_mask &= season_mask
            away_mask &= season_mask

        if venue == "home":
            away_mask = pd.Series([False] * len(df), index=df.index)
        elif venue == "away":
            home_mask = pd.Series([False] * len(df), index=df.index)

        home_games = df[home_mask]
        away_games = df[away_mask]

        # Vectorised result calculation
        hw = int((home_games["home_goal"] > home_games["away_goal"]).sum())
        hd = int((home_games["home_goal"] == home_games["away_goal"]).sum())
        hl = int((home_games["home_goal"] < home_games["away_goal"]).sum())

        aw = int((away_games["away_goal"] > away_games["home_goal"]).sum())
        ad = int((away_games["away_goal"] == away_games["home_goal"]).sum())
        al = int((away_games["away_goal"] < away_games["home_goal"]).sum())

        wins = hw + aw
        draws = hd + ad
        losses = hl + al
        goals_for = int(home_games["home_goal"].sum() + away_games["away_goal"].sum())
        goals_against = int(
            home_games["away_goal"].sum() + away_games["home_goal"].sum()
        )
        total = wins + draws + losses

        return {
            "team": team,
            "matches": total,
            "wins": wins,
            "draws": draws,
            "losses": losses,
            "goals_for": goals_for,
            "goals_against": goals_against,
            "goal_diff": goals_for - goals_against,
            "win_rate": round(wins / total * 100, 1) if total > 0 else 0.0,
            "competition": competition or "all",
            "season": season or "all",
            "venue": venue,
        }

    # ------------------------------------------------------------------ #
    # Player queries                                                       #
    # ------------------------------------------------------------------ #

    def search_players(
        self,
        name: Optional[str] = None,
        nationality: Optional[str] = None,
        club: Optional[str] = None,
        position: Optional[str] = None,
        min_overall: Optional[int] = None,
        limit: int = 20,
    ) -> list[dict]:
        """Search FIFA player data with any combination of filters."""
        df = self.loader.players
        mask = pd.Series([True] * len(df), index=df.index)

        if name:
            mask &= df["Name"].str.contains(name, case=False, na=False)
        if nationality:
            mask &= df["Nationality"].str.contains(nationality, case=False, na=False)
        if club:
            mask &= df["Club"].str.contains(club, case=False, na=False)
        if position:
            mask &= df["Position"].str.contains(position, case=False, na=False)
        if min_overall is not None:
            overall_num = pd.to_numeric(df["Overall"], errors="coerce")
            mask &= overall_num >= int(min_overall)

        result = df[mask].copy()
        result["_overall_num"] = pd.to_numeric(result["Overall"], errors="coerce")
        result = result.sort_values("_overall_num", ascending=False).head(limit)

        cols = ["Name", "Age", "Nationality", "Overall", "Potential", "Club", "Position"]
        available = [c for c in cols if c in result.columns]
        return result[available].to_dict("records")

    def players_by_club_summary(self, nationality: str = "Brazil") -> list[dict]:
        """
        Return clubs with a count and average rating for players of the given nationality,
        sorted by player count descending.
        """
        df = self.loader.players
        mask = df["Nationality"].str.contains(nationality, case=False, na=False)
        filtered = df[mask].copy()
        filtered["_overall_num"] = pd.to_numeric(filtered["Overall"], errors="coerce")

        summary = (
            filtered.groupby("Club")
            .agg(count=("Name", "count"), avg_rating=("_overall_num", "mean"))
            .reset_index()
            .sort_values("count", ascending=False)
        )
        return [
            {
                "club": row["Club"],
                "player_count": int(row["count"]),
                "avg_rating": round(float(row["avg_rating"]), 1),
            }
            for _, row in summary.iterrows()
        ]

    # ------------------------------------------------------------------ #
    # Competition standings                                                #
    # ------------------------------------------------------------------ #

    def get_standings(
        self,
        competition: str,
        season: int,
        source_filter: Optional[str] = None,
    ) -> list[dict]:
        """
        Calculate and return standings for a competition/season pair.

        Points: Win = 3, Draw = 1, Loss = 0.
        Tiebreaker: goal difference, then goals scored.
        """
        df = self.loader.matches
        comp = _resolve_competition(competition)
        mask = _comp_mask(df["competition"], comp)
        mask &= df["season"] == int(season)

        if source_filter:
            mask &= df["source"] == source_filter

        season_df = df[mask]
        if season_df.empty:
            return []

        # Collect all unique teams (normalized)
        teams = sorted(
            set(season_df["home_team"].unique()) | set(season_df["away_team"].unique())
        )

        standings: list[dict] = []
        for team in teams:
            hg = season_df[season_df["home_team"] == team]
            ag = season_df[season_df["away_team"] == team]

            hw = int((hg["home_goal"] > hg["away_goal"]).sum())
            hd = int((hg["home_goal"] == hg["away_goal"]).sum())
            hl = int((hg["home_goal"] < hg["away_goal"]).sum())
            hgf = int(hg["home_goal"].sum())
            hga = int(hg["away_goal"].sum())

            aw = int((ag["away_goal"] > ag["home_goal"]).sum())
            ad = int((ag["away_goal"] == ag["home_goal"]).sum())
            al = int((ag["away_goal"] < ag["home_goal"]).sum())
            agf = int(ag["away_goal"].sum())
            aga = int(ag["home_goal"].sum())

            w, d, l = hw + aw, hd + ad, hl + al
            gf, ga = hgf + agf, hga + aga
            pts = w * 3 + d

            # Get a human-readable name: prefer the raw name most commonly used
            raw_home = hg["home_team_raw"].mode()
            raw_away = ag["away_team_raw"].mode()
            if not raw_home.empty:
                display = raw_home.iloc[0]
            elif not raw_away.empty:
                display = raw_away.iloc[0]
            else:
                display = team

            standings.append(
                {
                    "team": display,
                    "played": w + d + l,
                    "wins": w,
                    "draws": d,
                    "losses": l,
                    "goals_for": gf,
                    "goals_against": ga,
                    "goal_diff": gf - ga,
                    "points": pts,
                }
            )

        return sorted(
            standings,
            key=lambda x: (-x["points"], -x["goal_diff"], -x["goals_for"]),
        )

    # ------------------------------------------------------------------ #
    # Statistical analysis                                                 #
    # ------------------------------------------------------------------ #

    def get_global_stats(
        self,
        competition: Optional[str] = None,
        season: Optional[int] = None,
    ) -> dict:
        """Return aggregated statistics: goals per match, home win rate, etc."""
        df = self.loader.matches
        mask = pd.Series([True] * len(df), index=df.index)

        if competition:
            comp = _resolve_competition(competition)
            mask &= _comp_mask(df["competition"], comp)
        if season is not None:
            mask &= df["season"] == int(season)

        subset = df[mask]
        total = len(subset)
        if total == 0:
            return {"error": "No matches found for given filters."}

        total_goals = int((subset["home_goal"] + subset["away_goal"]).sum())
        home_wins = int((subset["home_goal"] > subset["away_goal"]).sum())
        away_wins = int((subset["away_goal"] > subset["home_goal"]).sum())
        draws = int((subset["home_goal"] == subset["away_goal"]).sum())

        return {
            "total_matches": total,
            "total_goals": total_goals,
            "avg_goals_per_match": round(total_goals / total, 2),
            "home_wins": home_wins,
            "home_win_rate": round(home_wins / total * 100, 1),
            "away_wins": away_wins,
            "away_win_rate": round(away_wins / total * 100, 1),
            "draws": draws,
            "draw_rate": round(draws / total * 100, 1),
            "competition": competition or "all",
            "season": season or "all",
        }

    def biggest_wins(
        self,
        competition: Optional[str] = None,
        season: Optional[int] = None,
        limit: int = 10,
    ) -> list[dict]:
        """Return matches with the largest goal-difference margin."""
        df = self.loader.matches
        mask = pd.Series([True] * len(df), index=df.index)

        if competition:
            comp = _resolve_competition(competition)
            mask &= _comp_mask(df["competition"], comp)
        if season is not None:
            mask &= df["season"] == int(season)

        subset = df[mask].copy()
        subset["margin"] = (subset["home_goal"] - subset["away_goal"]).abs()
        result = subset.nlargest(limit, "margin")
        return result.to_dict("records")

    def top_scoring_teams(
        self,
        competition: Optional[str] = None,
        season: Optional[int] = None,
        limit: int = 10,
    ) -> list[dict]:
        """Return teams ranked by total goals scored."""
        df = self.loader.matches
        mask = pd.Series([True] * len(df), index=df.index)

        if competition:
            comp = _resolve_competition(competition)
            mask &= _comp_mask(df["competition"], comp)
        if season is not None:
            mask &= df["season"] == int(season)

        subset = df[mask]

        # Sum goals as home team + goals as away team
        home_goals = subset.groupby("home_team")["home_goal"].sum()
        away_goals = subset.groupby("away_team")["away_goal"].sum()
        total = home_goals.add(away_goals, fill_value=0).sort_values(ascending=False)

        results = []
        for team_norm, goals in total.head(limit).items():
            # Pick a display name
            raw = subset[subset["home_team"] == team_norm]["home_team_raw"]
            if raw.empty:
                raw = subset[subset["away_team"] == team_norm]["away_team_raw"]
            display = raw.mode().iloc[0] if not raw.empty else team_norm
            results.append({"team": display, "goals_scored": int(goals)})

        return results

File: experiments-local/experiment-mu-sonnet-claudecode/runs/agent=claude-code_language=typescript_model=sonnet-4.6_tooling=none/rep1/src/query-engine.ts
import { normalizeTeam, teamMatches } from "./data-loader.js";

function normalizeStr(s: string): string {
  return s.toLowerCase().normalize("NFD").replace(/[̀-ͯ]/g, "");
}
import type {
  NormalizedMatch,
  FifaPlayer,
  ExtendedMatch,
  TeamStats,
} from "./types.js";

export interface Database {
  matches: NormalizedMatch[];
  extended: ExtendedMatch[];
  players: FifaPlayer[];
}

// --- Match queries ---

export function searchMatches(
  db: Database,
  opts: {
    team?: string;
    homeTeam?: string;
    awayTeam?: string;
    opponent?: string;
    season?: number;
    competition?: string;
    dateFrom?: string;
    dateTo?: string;
    limit?: number;
  }
): NormalizedMatch[] {
  let results = db.matches;

  if (opts.homeTeam) {
    const ht = normalizeTeam(opts.homeTeam);
    results = results.filter((m) => teamMatches(ht, m.home_team));
  }

  if (opts.awayTeam) {
    const at = normalizeTeam(opts.awayTeam);
    results = results.filter((m) => teamMatches(at, m.away_team));
  }

  if (opts.team && !opts.homeTeam && !opts.awayTeam) {
    const t = normalizeTeam(opts.team);
    results = results.filter(
      (m) => teamMatches(t, m.home_team) || teamMatches(t, m.away_team)
    );
  }

  if (opts.opponent && opts.team) {
    const t = normalizeTeam(opts.team);
    const o = normalizeTeam(opts.opponent);
    results = results.filter((m) => {
      const homeIsTeam = teamMatches(t, m.home_team);
      const awayIsTeam = teamMatches(t, m.away_team);
      const homeIsOpp = teamMatches(o, m.home_team);
      const awayIsOpp = teamMatches(o, m.away_team);
      return (homeIsTeam && awayIsOpp) || (awayIsTeam && homeIsOpp);
    });
  } else if (opts.opponent && !opts.team) {
    const o = normalizeTeam(opts.opponent);
    results = results.filter(
      (m) => teamMatches(o, m.home_team) || teamMatches(o, m.away_team)
    );
  }

  if (opts.season) {
    results = results.filter((m) => m.season === opts.season);
  }

  if (opts.competition) {
    const comp = normalizeStr(opts.competition);
    results = results.filter((m) =>
      normalizeStr(m.competition).includes(comp)
    );
  }

  if (opts.dateFrom) {
    results = results.filter((m) => m.date >= opts.dateFrom!);
  }
  if (opts.dateTo) {
    results = results.filter((m) => m.date <= opts.dateTo!);
  }

  // Sort by date descending
  results.sort((a, b) => b.date.localeCompare(a.date));

  if (opts.limit) {
    results = results.slice(0, opts.limit);
  }

  return results;
}

// --- Team stats ---

export function getTeamStats(
  db: Database,
  team: string,
  opts: { season?: number; competition?: string; homeOnly?: boolean; awayOnly?: boolean }
): TeamStats {
  const t = normalizeTeam(team);
  let matches = db.matches;

  if (opts.season) matches = matches.filter((m) => m.season === opts.season);
  if (opts.competition) {
    const comp = normalizeStr(opts.competition);
    matches = matches.filter((m) => normalizeStr(m.competition).includes(comp));
  }

  const stats: TeamStats = {
    team,
    matches: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goals_for: 0,
    goals_against: 0,
    goal_difference: 0,
    points: 0,
  };

  for (const m of matches) {
    const isHome = teamMatches(t, m.home_team);
    const isAway = teamMatches(t, m.away_team);

    if (!isHome && !isAway) continue;
    if (opts.homeOnly && !isHome) continue;
    if (opts.awayOnly && !isAway) continue;

    stats.matches++;
    const gf = isHome ? m.home_goal : m.away_goal;
    const ga = isHome ? m.away_goal : m.home_goal;
    stats.goals_for += gf;
    stats.goals_against += ga;

    if (gf > ga) {
      stats.wins++;
      stats.points += 3;
    } else if (gf === ga) {
      stats.draws++;
      stats.points += 1;
    } else {
      stats.losses++;
    }
  }

  stats.goal_difference = stats.goals_for - stats.goals_against;
  return stats;
}

// --- Head-to-head ---

export function headToHead(
  db: Database,
  team1: string,
  team2: string,
  opts: { season?: number; competition?: string; limit?: number }
): {
  matches: NormalizedMatch[];
  team1_wins: number;
  team2_wins: number;
  draws: number;
  team1_goals: number;
  team2_goals: number;
} {
  const t1 = normalizeTeam(team1);
  const t2 = normalizeTeam(team2);

  let matches = db.matches.filter((m) => {
    const t1Home = teamMatches(t1, m.home_team);
    const t1Away = teamMatches(t1, m.away_team);
    const t2Home = teamMatches(t2, m.home_team);
    const t2Away = teamMatches(t2, m.away_team);
    return (t1Home && t2Away) || (t1Away && t2Home);
  });

  if (opts.season) matches = matches.filter((m) => m.season === opts.season);
  if (opts.competition) {
    const comp = normalizeStr(opts.competition);
    matches = matches.filter((m) => normalizeStr(m.competition).includes(comp));
  }

  matches.sort((a, b) => b.date.localeCompare(a.date));

  let t1_wins = 0, t2_wins = 0, draws = 0;
  let t1_goals = 0, t2_goals = 0;

  for (const m of matches) {
    const t1IsHome = teamMatches(t1, m.home_team);
    const t1gf = t1IsHome ? m.home_goal : m.away_goal;
    const t2gf = t1IsHome ? m.away_goal : m.home_goal;
    t1_goals += t1gf;
    t2_goals += t2gf;
    if (t1gf > t2gf) t1_wins++;
    else if (t1gf < t2gf) t2_wins++;
    else draws++;
  }

  const limited = opts.limit ? matches.slice(0, opts.limit) : matches;
  return { matches: limited, team1_wins: t1_wins, team2_wins: t2_wins, draws, team1_goals: t1_goals, team2_goals: t2_goals };
}

// --- Standings ---

export function getStandings(
  db: Database,
  season: number,
  competition?: string
): TeamStats[] {
  const teamsMap = new Map<string, TeamStats>();

  let matches = db.matches.filter((m) => m.season === season);
  if (competition) {
    const comp = normalizeStr(competition);
    matches = matches.filter((m) => normalizeStr(m.competition).includes(comp));
  }

  function getOrCreate(name: string): TeamStats {
    const key = normalizeTeam(name);
    if (!teamsMap.has(key)) {
      teamsMap.set(key, {
        team: name,
        matches: 0, wins: 0, draws: 0, losses: 0,
        goals_for: 0, goals_against: 0, goal_difference: 0, points: 0,
      });
    }
    return teamsMap.get(key)!;
  }

  for (const m of matches) {
    const home = getOrCreate(m.home_team);
    const away = getOrCreate(m.away_team);

    home.matches++; away.matches++;
    home.goals_for += m.home_goal; home.goals_against += m.away_goal;
    away.goals_for += m.away_goal; away.goals_against += m.home_goal;

    if (m.home_goal > m.away_goal) {
      home.wins++; home.points += 3;
      away.losses++;
    } else if (m.home_goal < m.away_goal) {
      away.wins++; away.points += 3;
      home.losses++;
    } else {
      home.draws++; home.points += 1;
      away.draws++; away.points += 1;
    }
  }

  const standings = Array.from(teamsMap.values()).map((s) => ({
    ...s,
    goal_difference: s.goals_for - s.goals_against,
  }));

  // Sort by points, then goal difference, then goals for
  standings.sort((a, b) =>
    b.points - a.points ||
    b.goal_difference - a.goal_difference ||
    b.goals_for - a.goals_for
  );

  return standings;
}

// --- Player queries ---

export function searchPlayers(
  db: Database,
  opts: {
    name?: string;
    nationality?: string;
    club?: string;
    position?: string;
    minOverall?: number;
    maxAge?: number;
    limit?: number;
  }
): FifaPlayer[] {
  let players = db.players;

  if (opts.name) {
    const n = opts.name.toLowerCase().normalize("NFD").replace(/[̀-ͯ]/g, "");
    players = players.filter((p) =>
      p.Name.toLowerCase().normalize("NFD").replace(/[̀-ͯ]/g, "").includes(n)
    );
  }

  if (opts.nationality) {
    const nat = opts.nationality.toLowerCase();
    players = players.filter((p) => p.Nationality.toLowerCase().includes(nat));
  }

  if (opts.club) {
    const club = opts.club.toLowerCase().normalize("NFD").replace(/[̀-ͯ]/g, "");
    players = players.filter((p) =>
      p.Club.toLowerCase().normalize("NFD").replace(/[̀-ͯ]/g, "").includes(club)
    );
  }

  if (opts.position) {
    const pos = opts.position.toUpperCase();
    players = players.filter((p) => p.Position.toUpperCase().includes(pos));
  }

  if (opts.minOverall !== undefined) {
    players = players.filter((p) => p.Overall >= opts.minOverall!);
  }

  if (opts.maxAge !== undefined) {
    players = players.filter((p) => p.Age <= opts.maxAge!);
  }

  players.sort((a, b) => b.Overall - a.Overall);

  if (opts.limit) players = players.slice(0, opts.limit);

  return players;
}

// --- Statistics ---

export function getGlobalStats(db: Database, competition?: string): {
  total_matches: number;
  total_goals: number;
  avg_goals_per_match: number;
  home_wins: number;
  away_wins: number;
  draws: number;
  home_win_rate: number;
} {
  let matches = db.matches;
  if (competition) {
    const comp = normalizeStr(competition);
    matches = matches.filter((m) => normalizeStr(m.competition).includes(comp));
  }

  let total_goals = 0, home_wins = 0, away_wins = 0, draws = 0;
  for (const m of matches) {
    total_goals += m.home_goal + m.away_goal;
    if (m.home_goal > m.away_goal) home_wins++;
    else if (m.home_goal < m.away_goal) away_wins++;
    else draws++;
  }

  const total_matches = matches.length;
  return {
    total_matches,
    total_goals,
    avg_goals_per_match: total_matches > 0 ? +(total_goals / total_matches).toFixed(2) : 0,
    home_wins,
    away_wins,
    draws,
    home_win_rate: total_matches > 0 ? +((home_wins / total_matches) * 100).toFixed(1) : 0,
  };
}

export function getBiggestWins(
  db: Database,
  limit = 10,
  competition?: string
): Array<NormalizedMatch & { margin: number }> {
  let matches = db.matches;
  if (competition) {
    const comp = normalizeStr(competition);
    matches = matches.filter((m) => normalizeStr(m.competition).includes(comp));
  }

  const withMargin = matches.map((m) => ({
    ...m,
    margin: Math.abs(m.home_goal - m.away_goal),
  }));

  withMargin.sort((a, b) => b.margin - a.margin || b.home_goal + b.away_goal - (a.home_goal + a.away_goal));
  return withMargin.slice(0, limit);
}

export function getExtendedStats(
  db: Database,
  team: string,
  opts: { limit?: number } = {}
): ExtendedMatch[] {
  const t = normalizeTeam(team);
  let results = db.extended.filter(
    (m) => teamMatches(t, m.home) || teamMatches(t, m.away)
  );
  if (opts.limit) results = results.slice(0, opts.limit);
  return results;
}
